File: Days.32/Python/5.Largest-Area-Histogram.py
# Largest Area Histogram using Stack
# Time Complexity   : O(n)
# Space Complexity  : O(n)


def LargestAreaHistogram(histogram,n):
    stack=[]
    
    #Next Smaller Left
    nsl=[-1 for i in range(n)]
    for i in range(n):
        while len(stack)>0 and histogram[i]<=histogram[stack[-1]]:
            stack.pop()
        if len(stack)==0:
            nsl[i]=-1
        else:
            nsl[i]=stack[-1]
        stack.append(i)
    
    while len(stack)>0:
        stack.pop()
    
    #Next Smaller Right
    nsr=[n for i in range(n)]
    for i in range(n-1,-1,-1):
        while len(stack)>0 and histogram[i]<=histogram[stack[-1]]:
            stack.pop()
        if len(stack)==0:
            nsr[i]=n
        else:
            nsr[i]=stack[-1]
        stack.append(i)
    maxarea=-2**63
    for i in range(n):
        maxarea=max(maxarea,(nsr[i]-nsl[i]-1)*histogram[i]);
    return maxarea

def main():
    n=int(input())
    histogram=list(map(int,input().split()))
    maxarea=LargestAreaHistogram(histogram,n)
    print(maxarea)



main()

File: Days.32/C++/5.Largest-Area-Histogram.cpp
// Largest Area Histogram using Stack
// Time Complexity  : O(n)
// Space Complexity : O(n)

#include<iostream>
#include<vector>
#include<climits>
#include<stack>
using namespace std;


int LargestAreaHistogram(vector<int> &histogram,int &n)
{
    stack<int> st;
    vector<int> nsl(n,-1);
    vector<int> nsr(n,n);
    
    // Next Smaller Right
    for (int i=n-1; i>=0; i--)
    {
        while (!st.empty() and histogram[i]<=histogram[st.top()])
        {
            st.pop();
        }
        if (st.empty())
        {
            nsr[i]=n;
        }
        else
        {
            nsr[i]=st.top();
        }
        st.push(i);
    }
    
    while (!st.empty())
    {
        st.pop();
    }
    
    // Next Smaller Left
    for(int i=0; i<n; i++)
    {
        while (!st.empty() and histogram[i]<=histogram[st.top()])
        {
            st.pop();
        }
        if(st.empty())
        {
            nsl[i]=-1;
        }
        else
        {
            nsl[i]=st.top();
        }
        st.push(i);
    }
    
    int maxarea=INT_MIN;
    for(int i=0; i<n; i++)
    {
        maxarea=max(maxarea,(nsr[i]-nsl[i]-1)*histogram[i]);
    }
    return maxarea;
}


int main()
{
    int n;
    cin>>n;
    vector<int> v(n);
    for(int i=0; i<n; i++)
    {
        cin>>v[i];
    }
    int maxarea=LargestAreaHistogram(v,n);
    cout<<maxarea<<"\n";
    return 0;
}